packages/block-library/src/button/index.php

```diff
@@ -0,0 +1,79 @@
+<?php
+/**
+ * Server-side rendering of the `core/button` block.
+ *
+ * @package WordPress
+ */
+
+ /**
+ * Renders the `core/button` block on the server,
+ *
+ * @since 6.6.0
+ *
+ * @param array    $attributes The block attributes.
+ * @param string   $content    The block content.
+ * @param WP_Block $block      The block object.
+ *
+ * @return string The block content.
+ */
+function render_block_core_button( $attributes, $content, $block ) {
+	$p = new WP_HTML_Tag_Processor( $content );
+
+	/*
+	 * The button block can render an `<a>` or `<button>` and also has a
+	 * `<div>` wrapper. Find the a or button tag.
+	 *
+	 */
+	$tag = null;
+	while ( $p->next_tag() ) {
+		$tag = $p->get_tag();
+		if ( 'A' === $tag || 'BUTTON' === $tag ) {
+			break;
+		}
+	}
+
+	// If this happens, the likelihood is there's no block content.
+	if ( null === $tag ) {
+		return '';
+	}
+
+	// Build a string of the inner text.
+	$text = '';
+	while ( $p->next_token() ) {
+		switch ( $p->get_token_name() ) {
+			case '#text':
+				$text .= $p->get_modifiable_text();
+				break;
+
+			case 'BR':
+				$text .= '';
+				break;
+		}
+	}
```

Comment:
I haven't written a lot of tag processor code, so would be happy to receive feedback on this.

Reply:
this is great @talldan! here are some thoughts. it _looks_ like you're attempting to remove all content inside the button except for the plain text content. that is, no formatting, no spans, no inline images, etc… I don't speak to whether that's a great idea or not, but I can provide feedback for the HTML API use in doing so.

first of all, this is technically the domain of the HTML Processor, which knows HTML structure, but since that's not really ready yet this will suffice. you did the right thing using the Tag Processor.

this code, however, is going to grab only the first text nodeat the start of the `A` or `BUTTON`. Maybe we don't expect any other content, but if you want to (unreliably) limit it to the inner contents, then you can add one more step which would improve the reliability a little bit and capture more text.

```php
// Build a string of the inner text.
$text = '';
while ( $p->next_token() ) {
	switch ( $p->get_token_name() ) {
		/*
         * Neither `A` nor `BUTTON` can nest, so it doesn't matter
         * if these are opening or closing tags. Both will close
         * the open element.
         */
		case $tag:
			break;

		case '#text':
			$text .= $p->get_modifiable_text();
			continue;

		case 'BR':
			$text .= "\n";
			continue;
	}
}
```

Reply:
> case $tag:
>	break;

Thanks for the feedback! Should this also break out of the while loop in addition to the switch? I believe in PHP `break 2` can be used to break two levels, but I might change it from a `switch` to an `if` statement.

Reply:
> it looks like you're attempting to remove all content inside the button except for the plain text content. that is, no formatting, no spans, no inline images, etc… I don't speak to whether that's a great idea or not, but I can provide feedback for the HTML API use in doing so.

The goal in this PR is for the replace the old static implementation (removed in this PR):
https://github.com/WordPress/gutenberg/blob/dbf201449e9736f672b61e422787d47659db327a/packages/block-library/src/button/save.js#L34-L36

Retaining the HTML from the save function allows for bindings to work with empty buttons.

It looks like the conditional logic introduced as a result of this issue - https://github.com/WordPress/gutenberg/issues/17221. 

It's not an accessibility reason. How many years have I incorrectly believed that?!  I'll update the comment in the code.

Reply:
I did some further testing and realized there's also the case of inline images. I've now simplified the code down to just this:
```php
// If the next token is the closing tag, the button is empty.
$p->next_token();
$is_empty = $p->get_token_name() === $tag;

if ( $is_empty ) {
	return '';
}
```

Which should better match how `RichText.isEmpty` works (even a space is considered content) - https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/is-empty.js.

I should've endeavoured to match that implementation first.

Anything I might be overlooking here?

Reply:
It's probably good enough, but keep in mind that there could be comments in there too.

```php
$is_empty = true;
while ( $p->next_token() && $tag !== $p->get_token_name() && $is_empty ) {
	if ( '#comment' !== $p->get_token_type() ) {
		// Anything else implies this is not empty.
		$is_empty = false;
	}
}
```